decoder/body_extensions_test.go

```diff
@@ -460,6 +460,452 @@ variable "test" {
+								Byte:   45,
+							},
+						},
+						NewText: "var.test",
+						Snippet: "var.test",
+					},
+					Kind: lang.TraversalCandidateKind,
+				},
+				{
+					Label:  "[ string ]",
+					Detail: "set of string",
+					TextEdit: lang.TextEdit{
+						Range: hcl.Range{
+							Filename: "test.tf",
+							Start: hcl.Pos{
+								Line:   2,
+								Column: 14,
+								Byte:   45,
+							},
+							End: hcl.Pos{
+								Line:   2,
+								Column: 14,
+								Byte:   45,
+							},
+						},
+						NewText: `[ "" ]`,
+						Snippet: `[ "${1:value}" ]`,
+					},
+					Kind: lang.SetCandidateKind,
+				},
+			}),
+		},
+		{
+			"for_each attribute completion with DependentBody",
+			&schema.BodySchema{
+				Blocks: map[string]*schema.BlockSchema{
+					"resource": {
+						Labels: []*schema.LabelSchema{
+							{
+								Name:     "type",
+								IsDepKey: true,
+							},
+							{
+								Name: "name",
+							},
+						},
+						Body: &schema.BodySchema{
+							Extensions: &schema.BodyExtensions{
+								ForEach: true,
+							},
+						},
+						DependentBody: map[schema.SchemaKey]*schema.BodySchema{
+							schema.NewSchemaKey(schema.DependencyKeys{
+								Labels: []schema.LabelDependent{
+									{
+										Index: 0,
+										Value: "aws_instance",
+									},
+								},
+							}): {
+								Attributes: map[string]*schema.AttributeSchema{
+									"type": {
+										Expr:       schema.LiteralTypeOnly(cty.String),
+										IsOptional: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			reference.Targets{},
+			`resource "aws_instance" "foo" {
+
+}`,
+			hcl.Pos{
+				Line:   2,
+				Column: 1,
+				Byte:   32,
+			},
+			lang.CompleteCandidates([]lang.Candidate{
+				{
+					Label: "for_each",
+					Description: lang.MarkupContent{
+						Value: "A meta-argument that accepts a map or a set of strings, ",
+						Kind:  lang.MarkdownKind,
+					},
+					Detail: "optional, map of string or set of string",
+					TextEdit: lang.TextEdit{
+						Range: hcl.Range{
+							Filename: "test.tf",
+							Start: hcl.Pos{
+								Line:   2,
+								Column: 1,
+								Byte:   32,
+							},
+							End: hcl.Pos{
+								Line:   2,
+								Column: 1,
+								Byte:   32,
+							},
+						},
+						NewText: "for_each",
+						Snippet: "for_each = ",
+					},
+					Kind:           lang.AttributeCandidateKind,
+					TriggerSuggest: true,
+				},
+				{
+					Label:  "type",
+					Detail: "optional, string",
+					TextEdit: lang.TextEdit{
+						Range: hcl.Range{
+							Filename: "test.tf",
+							Start: hcl.Pos{
+								Line:   2,
+								Column: 1,
+								Byte:   32,
+							},
+							End: hcl.Pos{
+								Line:   2,
+								Column: 1,
+								Byte:   32,
+							},
+						},
+						NewText: "type",
+						Snippet: `type = "${1:value}"`,
+					},
+					Kind: lang.AttributeCandidateKind,
+				},
+			}),
+		},
+		{
+			"for_each does not complete with count present",
```

Comment:
This is a useful test, but it would be also nice to validate it earlier, before the [invalid] schema even gets to the decoder, here: https://github.com/hashicorp/hcl-lang/blob/e12fb9092dc7bde532c1b1346069a23d9eb2c7f9/schema/body_schema.go#L154-L181

This validation provides some form of basic "integrity checks" and it is something we typically run as part of tests within `terraform-schema`.

It doesn't negate the need to catch it here also, but it can be an additional early warning.